python/tk_framework_desktopserver/shotgun/scripts/cache_commands.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sys
+import logging
+import cPickle
+
+class _BootstrapLogHandler(logging.StreamHandler):
+    """
+    Manually flushes emitted records for js to pickup.
+    """
+
+    def emit(self, record):
+        """
+        Forwards the record back to to js via the engine communicator.
+
+        :param record: The record to log.
+        """
+        super(_BootstrapLogHandler, self).emit(record)
+
+        # always flush to ensure its seen by the js process
+        self.flush()
+
+def get_sgtk_logger(sgtk):
+    """
+    Sets up a log handler and logger.
+
+    :param sgtk: An sgtk module reference.
+
+    :returns: A logger and log handler.
+    """
+    # add a custom handler to the root logger so that all toolkit log messages
+    # are forwarded back to python via the communicator
+    bootstrap_log_formatter = logging.Formatter("[%(levelname)s]: %(message)s")
+    bootstrap_log_handler = _BootstrapLogHandler()
+    bootstrap_log_handler.setFormatter(bootstrap_log_formatter)
+    bootstrap_log_handler.setLevel(logging.DEBUG)
+
+    # now get a logger to use during bootstrap
+    sgtk_logger = sgtk.LogManager.get_logger("%s.%s" % ("tk-shotgun", "bootstrap"))
+    sgtk.LogManager().initialize_custom_handler(bootstrap_log_handler)
+
+    # allows for debugging to be turned on by the plugin build process
+    sgtk.LogManager().global_debug = True
+
+    # initializes the file where logging output will go
+    sgtk.LogManager().initialize_base_file_handler("tk-shotgun")
+    sgtk_logger.debug("Log dir: %s" % (sgtk.LogManager().log_folder))
+
+    return sgtk_logger, bootstrap_log_handler
+
+def cache(cache_file, data, base_configuration, hash_data, engine_name):
+    import sqlite3
+    import sgtk
+
+    entity = dict(type=data["entity_type"], id=data["entity_id"])
+    project_entity = dict(id=data["project_id"], type="Project")
+
+    # Setup the bootstrap manager.
+    toolkit_mgr = sgtk.bootstrap.ToolkitManager()
+    toolkit_mgr.plugin_id = "shotgun_toolkit_menu_caching"
```

Comment:
Why not use basic.shotgun? Somehow might want to manage their shotgun actions through a second config, just like you could manage your maya actions with a second config.